File: Frontend/public/javascripts/pick_and_place.js
async function start() {
    api.clearResult();
    log("Starting...");
    
    log("Getting Latest Scene Data...");
    scene_data = await api.getSceneData();
    log("Received Scene Data " + JSON.stringify(scene_data));

    log("Planning pick trajectory...")
    posePoint = scene_data.object_location;
    posePoint.z += 0.10; // stop just above object.
    posePoint.y -= 0.01;
    posePoint.x -= 0.01;
    poseOrientation = {x : -0.5, y : -0.5, z : 0.5, w : -0.5}; // perpendicular to table.
    pose = {position: posePoint, orientation: poseOrientation};
    plan = await api.getPlan(pose);

    log("Executing pick move with plan...");
    await api.doTrajectory(plan.trajectory);
    await api.waitNextResult();
    log("Done");

    log("Opening gripper...");
    await api.doOpenGripper();
    await api.waitNextResult();
    log("Done");

    log("Planning lowering of arm over object...");
    pose.position.z -= 0.04;
    plan = await api.getPlan(pose);

    log("Lowering arm over object...");
    await api.doTrajectory(plan.trajectory);
    await api.waitNextResult();
    log("Done");

    log("Closing gripper...");
    await api.doCloseGripper();
    await api.waitNextResult();
    log("Done");

    log("Planning place trajectory...")
    posePoint = scene_data.target_location;
    posePoint.z += 0.15; // stop just above target.
    poseOrientation = {x : -0.5, y : -0.5, z : 0.5, w : -0.5}; // perpendicular to table.
    pose = {position: posePoint, orientation: poseOrientation};
    plan = await api.getPlan(pose);


    log("Executing place move with plan...");
    await api.doTrajectory(plan.trajectory);
    await api.waitNextResult();
    log("Done");

    log("Opening gripper...");
    await api.doOpenGripper();
    await api.waitNextResult();
    log("Done");
}

File: Frontend/public/workspaces/python/pick_and_place.py
import asyncio

from api import RobotApi

# Quaternion peripindicular to the niryo table.
TABLE_PERPINDICULAR = {'x' : -0.5, 'y' : -0.5, 'z' : 0.5, 'w' : -0.5}

async def main():
    api = RobotApi()
    await api.Initialize()

    scene_data = await api.GetSceneData()

    print('Executing pick trajactory...')
    posePoint = scene_data['object_location']
    posePoint['z'] = posePoint['z'] + 0.10 # stop just above object.
    posePoint['y'] = posePoint['y'] - 0.01
    posePoint['x'] = posePoint['x'] - 0.01 
    goal = {'position': posePoint, 'orientation': TABLE_PERPINDICULAR}
    plan = await api.GetPlan(goal)
    await api.DoTrajectory(plan['trajectory'])
    print('---> Done')

    print("Opening gripper...")
    await api.DoOpenGripper()
    print('---> Done')
    
    print('Executing lower trajectory...')
    posePoint['z'] = posePoint['z'] - 0.04
    plan = await api.GetPlan(goal)
    await api.DoTrajectory(plan['trajectory'])
    print('---> Done')
    
    print("Closing gripper...")
    await api.DoCloseGripper()
    print('---> Done')
    
    print('Executing place trajectory...')
    posePoint = scene_data['target_location']
    posePoint['z'] = posePoint['z'] + 0.15 # Just above.
    goal = {'position': posePoint, 'orientation': TABLE_PERPINDICULAR}
    plan = await api.GetPlan(goal)
    await api.DoTrajectory(plan['trajectory'])
    print('---> Done')
     
    print("Opening gripper...")
    await api.DoOpenGripper()
    print('---> Done')


if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())